EOSIO/eos-vm
# Fix Buildkite Pipeline + Refactor + Improvements

I was asked to fix the Buildkite pipeline on this branch. The CI scripts were a rats nest of BASH. After hours and hours of errors that were, at best, tangential to my changes, I decided it would be faster to refactor the scripts entirely than to continue trying to edit them in-place.  
  
### Refactor
I was able to eliminate the `general.sh` "library" altogether, which served to load in a series of unnecessary variables that were only used in two places. Next, I reduced `docker-hash.sh` from 25 lines to just three (simple) lines, and `generate-base-images.sh` from 17 to 11 lines, also reducing complexity. Then I noticed that `build.sh` and `test.sh` use the same code to run docker, so I was able to pull all the container management code into one file, `docker.sh`, which can be used by `build.sh` or `test.sh` or anything else....it doesn't care what you are doing. All it does is run your commands in the correct container (including searching for, pulling, building base images; starting the container with the right mounts, paths, BUILDKITE variables, etc).  
  
### Improvements
- macOS is now building and running tests
  - We chose to disable them last summer due to Anka fleet instability at the time
- Test step will fail if there are zero tests
- Developers have more control over what steps are run
  - Added `SKIP_LINUX='true|false'`
  - Added `SKIP_MAC='true|false'`
- Buildkite job timeouts are fixed
  - Default timeouts are set to 1.5x the current longest step duration on any major branch
- Docker base-image container size-on-disk reduced substantially
- The working directory was changed from `/workdir` to `/eos-vm` to prevent conflicts if multiple products are built in the same container
- The dockerfile SHA-1 hash no longer includes the filename
- CI scripts can be run locally (for debugging)
  - Declare `IMAGE_TAG='amazonlinux-2|centos-7.7|ubuntu-16.04|ubuntu-18.04'` to build in docker, or leave it unset for a native build
  - Note that, if the base-image does not exist, it will attempt a `docker push` and that step will fail if you are not authenticated to the Docker registry
- The scripts/pipeline shows you what commands it is running, for easier debugging
- Added build artifacts to `.gitignore`

## Tested
- [Build 944](https://buildkite.com/EOSIO/eos-vm/builds/944) -- successful base-image build
  - The build itself failed due to bugs which have since been fixed
- [Build 957](https://buildkite.com/EOSIO/eos-vm/builds/957) -- `eos-vm:eosio-2.0.x`
- [Build 958](https://buildkite.com/EOSIO/eos-vm/builds/958) -- `eos-vm:eosio`
- [Build 959](https://buildkite.com/EOSIO/eos-vm/builds/959) -- `eos-vm:develop`
- [GitHub Actions Run 86954827](https://github.com/EOSIO/eos-vm/actions/runs/86954827) -- community builds
  
## See Also
- [Pull Request 178](https://github.com/EOSIO/eos-vm/pull/178) -- `eos-vm:eosio-2.0.x`
- [Pull Request 179](https://github.com/EOSIO/eos-vm/pull/179) -- `eos-vm:eosio`
- [Pull Request 180](https://github.com/EOSIO/eos-vm/pull/180) -- `eos-vm:develop`
